app/services/oauth_clients/connection_manager.rb

```diff
@@ -88,10 +91,57 @@ def code_to_token(code)
       service_result = request_new_token(authorization_code: code)
       return service_result unless service_result.success?
 
-      # Create a new OAuthClientToken from Rack::OAuth::AccessToken::Bearer and return
-      ServiceResult.success(
-        result: create_new_oauth_client_token(service_result.result)
+      # Check for existing OAuthClientToken and update,
+      # or create a new one from Rack::OAuth::AccessToken::Bearer
+      oauth_client_token = get_existing_token
+      if oauth_client_token.present?
+        update_oauth_client_token(oauth_client_token, service_result.result)
+      else
+        oauth_client_token = create_new_oauth_client_token(service_result.result)
+      end
+
+      ServiceResult.success(result: oauth_client_token)
+    end
+
+    # Called by StorageRepresenter to inquire about the status of the OAuth2
+    # authentication server.
+    # Returns :connected/:authorization_failed or :error for a general error.
+    # We have decided to distinguish between only these 3 cases, because the
+    # front-end (and a normal user) probably wouldn't know how to deal with
+    # other options.
+    def authorization_state
+      oauth_client_token = get_existing_token
+      return :failed_authorization unless oauth_client_token
+
+      # Check for user information. This is the cheapest Nextcloud call that requires
+      # valid authentication, so we use it for testing the validity of the Bearer token.
+      # curl -H "Authorization: Bearer MY_TOKEN" -X GET 'https://my.nextcloud.org/ocs/v1.php/cloud/user' \
+      #      -H "OCS-APIRequest: true" -H "Accept: application/json"
+      RestClient.get(
+        File.join(oauth_client.integration.host, AUTHORIZATION_CHECK_PATH),
+        {
+          'Authorization' => "Bearer #{oauth_client_token.access_token}",
+          'OCS-APIRequest' => "true",
+          'Accept' => "application/json"
+        }
       )
```

Comment:
This operation does a costly `http` request to a different server. I get that it should represent the current state but I think it worth to balance it with the cost of slow response time. If I understand the use case this will be employed in correctly, in a sense you are already caching, since the state of the storage is fetched first, and only after that has returned, will the frontend do the next request where the file links are fetched. That fetching of file links will use the same auth token as before. So in the time between that first request to get the state and the second request, the auth token will be cached without knowing whether it will still be valid. 

Based on this assessment, I would opt for two things:

* Cache the existing tokens for a bit by either
  * putting it in the request store to at least not repeat the check within the same request (e.g. whenever more than one storage representer representing the same storage is rendered)
  * putting it in `OpenProject::Cache` (i.e. memcached) for some time (e.g. 5 minutes) so that all workers of the server can access it within that time frame without needing to recheck. That way, moving through a work package list to take a look at the respective file links becomes less costly. And whenever a request to the storage succeeded , reset the clock on the cache. The only reason not to do it might be to not wanting the sensitive token to reside in yet another place. If that is an argument, introduce a `last_checked` column in the database to do the caching there.
* Handle failures to fetch the file link's live info (I have not checked whether that is done already or not) since even between the two requests, a token might get blocked or because of a simple connection problem. 

Because of the second point, I am wondering whether it makes sense to fetch the status of the storage before the call to file links. It might also suffice to just deal with the failures in the second request since that has to happen anyway.

Reply:
I just want to add to the last point, as this was exactly, what we two were talking about a lot last week. ;)

If we do only one request to fetch file links from a storage, we basically have no possibility in our current response models to embody both the storage auth state (is it connected, is it even reachable, etc.) and the individual file link permission states. This was the reason we decided to go with two separate requests for the beginning, not caring too much about slow responses here (no premature optimization). Just remember our collection models with linked state of the storage.

I'd also argue, that just because of the possibility, the second request can fail after the first to storage itself succeeded, this is no reason to merge those. you can have an error handling on every request. In addition, we do have the semantic need to fetch information of the storage, as we display the storage sections with own information. Just imagine, we would have a slow response for fetching file links (e.g fetching the live data for around 50 links), then you actually just wait seconds to even see the storage itself. The request to fetch the storage information is much less costly.